table-to-blt.mjs

```diff
@@ -0,0 +1,147 @@
+  // NOTE: The first column is the name of the voter.
+  let table = dom.window.document.querySelector("table")
+
+  // Ignore first column in determining number of candidates.
+  const nbcandidates = table.querySelector("tr").querySelectorAll('td,th').length - 1;
+  let rows = table.querySelectorAll("tr");
+
+  // candidates is the list of candidate names, taken from the top (zeroth) row of the table.
+  const candidates = Array.from(rows[0].querySelectorAll('th')).map(s => s.innerHTML).slice(1) ;
+    
+  // Create ballots
+  const ballots = generateBallots(dom, rows, candidates);
+  
+  // Start output in BLT format
+  // First row is number of candidates and number of seats
+  console.log(nbcandidates + " " + nbseats);
+  // List of ballots, each with weight "1" and ending with "0"
+  console.log(ballots.map(s => "1 " + s.join(" ") + " 0").join("\n"));
+  // Zero separator
+  console.log("0");
+  // Names of candidates, ignoring the first column.
+  for (let i = 1; i <= nbcandidates; i++) {
+    console.log("\"Candidate " + i + "\"");
+  }
+  // Election name
+  console.log("\"" + electionname + "\"");
+}
+
+function generateBallots(dom, rows, candidates) {
+    let ballots = [];
+    // Handle rows. Ignore top (zeroth) row since that is candidate names.
+    for (let i = 1; i < rows.length; i++) {
+        // Handle cells in the row.
+        let cells = rows[i].querySelectorAll('td');
+	let unranked = Array.from(cells).map(getVote);
+
+        // Per OpenSTV, ballot is invalid if there are duplicate rankings.
+        if (duplicateRankings(unranked)) {
+	   console.error("Ballot ignored (duplicate rankings): " + rows[i].querySelector('th').innerHTML);
+	   continue;
+	}
+
+        // Sort the candidates so the array of cells goes from top ranked (1) to lowest ranked.
+        let rankedcells = [...unranked].sort(sortByRank);
+
+        // Per OpenSTV, ballot is invalid if there are skips in rankings.
+        if (!noSkips(rankedcells, cells.length)) {
+	   console.error("Ballot ignored (skips in rankings): " + rows[i].querySelector('th').innerHTML);
+	   continue;
+	}
+
+	let ordered = [];
+        for (let j = 0; j < cells.length; j++) {
+	  // Unranked candidates are demoted in the list by giving them very high rank (Number.POSITIVE_INFINITY).
+	  // Remove these candidates from the generated ballot.
+	  if (rankedcells[j].rank != Number.POSITIVE_INFINITY) {
+  	     ordered.push(candidates.indexOf(rankedcells[j].candidate) + 1);
+	  }
+        }
+
+	ballots.push(ordered);
+    }
+    return(ballots);
+ }
+
+function getVote (cell) {
+  // WBS form generates "Ranked N" (most of the time). Remove "Ranked"
+  const re1 = /.*\s+([0123456789])$/ ;
+  const re2 = /\s*Unranked\s*/ ;
+  let title = cell.getAttribute('title');
+  if (title != null) {
+     return { candidate: title.split(':')[0], rank: title.split(':')[1].replace(re1, "$1").replace(re2, Number.POSITIVE_INFINITY) };
+
+  } else {
+     return ("");
```

Comment:
if this situation arises, this would make other code breaks (e.g. when trying to sortByRank); since I think we don't expect that situation to be valid, I would throw an exception
```js
throw new Error(`Cell ${cell.outerHTML} does not follow expected format`);
```